File: pb_option1_vision/src/test_hsv_with_morph.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <chrono>
#include <thread>

int main() {
    cv::VideoCapture cap(0);  // 0 是默认摄像头ID
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open camera." << std::endl;
        return -1;
    }

    // 降分辨率以优化性能
    cap.set(cv::CAP_PROP_FRAME_WIDTH, 320);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, 240);

    // 创建窗口和轨迹条
    cv::namedWindow("HSV Adjust");
    int h_l = 0, h_u = 30, s_l = 48, s_u = 255, v_l = 80, v_u = 255;
    cv::createTrackbar("H Lower", "HSV Adjust", &h_l, 180);
    cv::createTrackbar("H Upper", "HSV Adjust", &h_u, 180);
    cv::createTrackbar("S Lower", "HSV Adjust", &s_l, 255);
    cv::createTrackbar("S Upper", "HSV Adjust", &s_u, 255);
    cv::createTrackbar("V Lower", "HSV Adjust", &v_l, 255);
    cv::createTrackbar("V Upper", "HSV Adjust", &v_u, 255);

    // 形状阈值
    double min_area = 5000.0;
    double min_circularity = 0.2;
    double aspect_min = 0.5;
    double aspect_max = 2.0;

    // 帧率控制：1 FPS (1s一次)
    int delay = 1000;  // ms

    while (true) {
        auto start_frame = std::chrono::steady_clock::now();

        cv::Mat frame;
        cap >> frame;
        if (frame.empty()) break;

        // 可选：水平翻转像镜子（如果指向反了）
        // cv::flip(frame, frame, 1);  // 取消注释如果需要

        cv::Mat hsv;
        cv::cvtColor(frame, hsv, cv::COLOR_BGR2HSV);

        cv::Scalar lower(h_l, s_l, v_l);
        cv::Scalar upper(h_u, s_u, v_u);

        cv::Mat mask;
        cv::inRange(hsv, lower, upper, mask);

        cv::Mat kernel = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(3, 3));  // 小kernel优化
        cv::morphologyEx(mask, mask, cv::MORPH_OPEN, kernel);
        cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, kernel);

        std::vector<std::vector<cv::Point>> contours;
        cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

        std::string direction;  // 默认空
        if (!contours.empty()) {
            auto largest_it = std::max_element(contours.begin(), contours.end(),
                                               [](const auto& a, const auto& b) {
                                                   return cv::contourArea(a) < cv::contourArea(b);
                                               });
            std::vector<cv::Point> largest = *largest_it;

            double area = cv::contourArea(largest);
            if (area >= min_area) {
                double perimeter = cv::arcLength(largest, true);
                double circularity = (perimeter > 0) ? (4 * CV_PI * area / (perimeter * perimeter)) : 0;
                if (circularity >= min_circularity) {
                    cv::Rect bbox = cv::boundingRect(largest);
                    double aspect_ratio = static_cast<double>(bbox.width) / bbox.height;
                    if (aspect_ratio >= aspect_min && aspect_ratio <= aspect_max) {
                        // 计算凸包
                        std::vector<cv::Point> hull;
                        cv::convexHull(largest, hull);

                        // 找极端点
                        auto left_it = std::min_element(hull.begin(), hull.end(),
                                                        [](const cv::Point& a, const cv::Point& b) { return a.x < b.x; });
                        cv::Point extreme_left = *left_it;

                        auto right_it = std::max_element(hull.begin(), hull.end(),
                                                         [](const cv::Point& a, const cv::Point& b) { return a.x < b.x; });
                        cv::Point extreme_right = *right_it;

                        auto top_it = std::min_element(hull.begin(), hull.end(),
                                                       [](const cv::Point& a, const cv::Point& b) { return a.y < b.y; });
                        cv::Point extreme_top = *top_it;

                        auto bottom_it = std::max_element(hull.begin(), hull.end(),
                                                          [](const cv::Point& a, const cv::Point& b) { return a.y < b.y; });
                        cv::Point extreme_bottom = *bottom_it;

                        // 中心
                        int center_x = (extreme_left.x + extreme_right.x) / 2;
                        int center_y = (extreme_top.y + extreme_bottom.y) / 2;

                        // 判断指向（调整逻辑：如果左极端更远→左转，反之右转）
                        int left_dist = center_x - extreme_left.x;
                        int right_dist = extreme_right.x - center_x;
                        if (right_dist > left_dist) {
                            direction = "右转";
                        } else if (left_dist > right_dist) {
                            direction = "左转";
                        }

                        if (!direction.empty()) {
                            cv::drawContours(frame, {largest}, -1, cv::Scalar(0, 255, 0), 2);
                            cv::circle(frame, extreme_left, 5, cv::Scalar(255, 0, 0), -1);
                            cv::circle(frame, extreme_right, 5, cv::Scalar(0, 0, 255), -1);
                            cv::circle(frame, cv::Point(center_x, center_y), 5, cv::Scalar(0, 255, 255), -1);
                            cv::putText(frame, direction, cv::Point(bbox.x, bbox.y - 10),
                                        cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 0), 2);

                            std::cout << "手指向: " << direction << std::endl;
                        }
                    }
                }
            }
        }

        cv::imshow("Original with Detection", frame);
        cv::imshow("Mask", mask);

        auto end_frame = std::chrono::steady_clock::now();
        auto frame_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_frame - start_frame).count();
        std::cout << "本帧处理时间: " << frame_time << "ms" << std::endl;

        // 延时控制帧率
        int key = cv::waitKey(delay);
        if (key == 'q') break;
    }

    cap.release();
    cv::destroyAllWindows();
    return 0;
}

File: pb_option1_vision/src/test_hsv_with_morph.py
import cv2
import numpy as np
import time

def nothing(x):
    pass

cap = cv2.VideoCapture(0)

# 降低分辨率，减少计算量，提高流畅度
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 320)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 240)

cv2.namedWindow('Control')

# 颜色范围故意放宽（尤其是苹果的红色/橙红/偏黄都包含）
cv2.createTrackbar('H Lower', 'Control', 0, 180, nothing)
cv2.createTrackbar('H Upper', 'Control', 40, 180, nothing)   # 放宽到40，包含橙红
cv2.createTrackbar('S Lower', 'Control', 60, 255, nothing)   # 降低饱和度下限
cv2.createTrackbar('S Upper', 'Control', 255, 255, nothing)
cv2.createTrackbar('V Lower', 'Control', 40, 255, nothing)   # 降低亮度下限
cv2.createTrackbar('V Upper', 'Control', 255, 255, nothing)

last_process_time = 0
process_interval = 1.0  # 每 1 秒处理一次

while True:
    ret, frame = cap.read()
    if not ret:
        break

    current_time = time.time()

    # 只每秒处理一次
    if current_time - last_process_time < process_interval:
        cv2.imshow('Detection', frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
        continue

    last_process_time = current_time

    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

    h_l = cv2.getTrackbarPos('H Lower', 'Control')
    h_u = cv2.getTrackbarPos('H Upper', 'Control')
    s_l = cv2.getTrackbarPos('S Lower', 'Control')
    s_u = cv2.getTrackbarPos('S Upper', 'Control')
    v_l = cv2.getTrackbarPos('V Lower', 'Control')
    v_u = cv2.getTrackbarPos('V Upper', 'Control')

    lower = np.array([h_l, s_l, v_l])
    upper = np.array([h_u, s_u, v_u])

    mask = cv2.inRange(hsv, lower, upper)

    kernel = np.ones((5, 5), np.uint8)
    mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
    mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)

    contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    command = ""  # 最终要执行的命令

    if contours:
        largest = max(contours, key=cv2.contourArea)
        area = cv2.contourArea(largest)

        # 过滤太小的物体
        if area < 600:
            cv2.imshow('Detection', frame)
            if cv2.waitKey(1) & 0xFF == ord('q'):
                break
            continue

        perimeter = cv2.arcLength(largest, True)
        circularity = 4 * np.pi * area / (perimeter ** 2) if perimeter > 0 else 0

        x, y, w, h = cv2.boundingRect(largest)
        aspect = float(w) / h if h > 0 else 1.0

        # ========================
        #   核心判断逻辑（形态为主）
        # ========================

        # 苹果：圆度高 + 宽高比接近1
        if (circularity > 0.62 and          # 圆度放宽到0.62
            0.7 < aspect < 1.45 and         # 宽高比更宽松
            area > 600):                    # 面积要求稍高，避免误判
            command = "苹果 → 右转"
            color = (0, 255, 0)  # 绿色框

        # 香蕉：圆度低 + 宽高比很大
        elif (circularity < 0.40 and
              1.9 < aspect < 6.0 and
              area > 600):
            command = "香蕉 → 左转"
            color = (0, 165, 255)  # 橙色框（区分）

        if command:
            cv2.drawContours(frame, [largest], -1, color, 2)
            cv2.putText(frame, command, (x, y - 10),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.8, color, 2)
            print(command)

    cv2.imshow('Detection', frame)
    cv2.imshow('Mask', mask)  # 想调试掩码时打开

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()